### ISSUE TITLE
Mac OS , many Unity instance, can't start debugger

### ISSUE THREAD

**Issue by fjqingyou** [Unity-Technologies/vscode-unity-debug#96]
Hello!
My English is very bad, so , I'm upload a screenshots.

![wx20180629-102426](https://user-images.githubusercontent.com/31087168/42069975-0a0abad0-7b87-11e8-9a59-b9bee7395bf0.png)

Is there any other material I need to provide?

**Comment by miniwolf** [comment]
Which version of the unity debugger are you using? The launch.json looks like it is from before 2.5.0 try the newest version on https://forum.unity.com/threads/unity-debugger-extension-for-visual-studio-code-2-0-x-preview-release.526263/ and delete the launch.json file and recreate it when attaching the debugger.

**Comment by fjqingyou** [comment]
@miniwolf 

VSCode 1.24.1
Debugger for Unity 2.6.0
Mono Debug 0.15.8
OS : macOS Sierra 10.12.6

**Comment by fjqingyou** [comment]
launch.json is plugin auto create , i don't it body is Meaning, so i not change it

**Comment by miniwolf** [comment]
Could you try removing it and generating it again?

**Comment by fjqingyou** [comment]
@miniwolf 

Hello, i remote it  and plugin auto generating a new file ,but i get a new issue, can't start debugger too
![wx20180702-153001](https://user-images.githubusercontent.com/31087168/42150575-e0d21a90-7e0c-11e8-984e-d615a084aca5.png)

**Comment by fjqingyou** [comment]
My Unity Version is 5.6.3f1

**Comment by miniwolf** [comment]
Hmm.. It may be that 5.6 does not generate the `EditorInstance.json` file. Or maybe it is placed somewhere differently?

**Comment by fjqingyou** [comment]
@miniwolf 
i'm not found EditorInstance.json file in Library directory,

i using mac finder find this project directory not found it too.

What should I do?

bogon:Majong_V2 xuchenglin$ ls
Assembly-CSharp-Editor-firstpass.csproj Assembly-CSharp.csproj                  Majong.userprefs                        ProjectSettings
Assembly-CSharp-Editor.csproj           Assets                                  Majong_v2.sln                           Temp
Assembly-CSharp-firstpass.csproj        Library                                 Majong_v2.userprefs                     obj



bogon:Majong_V2 xuchenglin$ cd Library
bogon:Library xuchenglin$ ls
AnnotationManager                       EditorUserBuildSettings.asset           ScriptAssemblies                        p4plugin.log
AssetImportState                        EditorUserSettings.asset                ScriptMapper                            shadercompiler-UnityShaderCompiler0.log
AssetServerCacheV3                      InspectorExpandedItems.asset            ServerPreferences.plist                 shadercompiler-UnityShaderCompiler1.log
AssetVersioning.db                      LastBuild.buildreport                   ShaderCache                             shadercompiler-UnityShaderCompiler2.log
AtlasCache                              LastSceneManagerSetup.txt               ShaderCache.db                          shadercompiler-UnityShaderCompiler3.log
BuildPlayer.prefs                       LibraryFormatVersion.txt                assetDatabase3
BuildSettings.asset                     MonoManager.asset                       expandedItems
CurrentLayout.dwlt                      ProjectSettings.asset                   metadata



bogon:Library xuchenglin$ ls -a
.                                       BuildPlayer.prefs                       LibraryFormatVersion.txt                assetDatabase3
..                                      BuildSettings.asset                     MonoManager.asset                       expandedItems
.DS_Store                               CurrentLayout.dwlt                      ProjectSettings.asset                   metadata
AnnotationManager                       EditorUserBuildSettings.asset           ScriptAssemblies                        p4plugin.log
AssetImportState                        EditorUserSettings.asset                ScriptMapper                            shadercompiler-UnityShaderCompiler0.log
AssetServerCacheV3                      InspectorExpandedItems.asset            ServerPreferences.plist                 shadercompiler-UnityShaderCompiler1.log
AssetVersioning.db                      LastBuild.buildreport                   ShaderCache                             shadercompiler-UnityShaderCompiler2.log
AtlasCache                              LastSceneManagerSetup.txt               ShaderCache.db                          shadercompiler-UnityShaderCompiler3.log
bogon:Library xuchenglin$

**Comment by miniwolf** [comment]
If this is the case then I will have to print a warning and ignore the tag in launch.json. This however means that multiple instances of unity editor at the same time does not work for all versions.
I will make a new release, so it will at least not crash.

**Comment by fjqingyou** [comment]
Thank you!

**Comment by miniwolf** [comment]
Can you verify whether this works for you? https://github.com/Unity-Technologies/vscode-unity-debug/releases/download/Version-2.6.2/unity-debug-2.6.2.vsix

**Comment by fjqingyou** [comment]
@miniwolf 
Yes, but 2018-7-2 22:55:45 for now on my location time, go to work tomorrow try it. My Home PC is Windows, work pc is Mac OS!

**Comment by miniwolf** [comment]
No worries. Just write when you have tried it out.

**Comment by fjqingyou** [comment]
@miniwolf 
apply unity-debug-2.6.2.vsix , 
if only one Unity instance, start debugger success
if have many Unity instance retry get a error MessageBox

error while processing request 'launch' (exception: Could not find file "/Users/xuchenglin/Project/MyProject/Majong_v2/Library/EditorInstance.json")

**Comment by miniwolf** [comment]
Alright yeah that was my bad. I forgot to actually put the executable inside the vsix file. Here you go: https://github.com/Unity-Technologies/vscode-unity-debug/releases/download/Version-2.6.2/unity-debug-2.6.2.vsix

**Comment by fjqingyou** [comment]
@miniwolf 

if only one Unity instance, start debugger success
if have many Unity instance retry get a error MessageBox

![qq20180703-180431](https://user-images.githubusercontent.com/31087168/42213541-a49989c0-7eeb-11e8-95b7-141d1e640b43.png)

**Comment by fjqingyou** [comment]
@miniwolf 
can you foreach all process , find all unity instance process ,show to popup list view items, attach user selected  unity instance ?

Can you do this?

**Comment by miniwolf** [comment]
If you read the message it tells you how to do that through the command in the command palette. Go to View > Command Palette... and type Unity Attach Debugger then it will give you the list.

**Comment by fjqingyou** [comment]
@miniwolf 

Sorry , This is my fault!

i'm get a new lssue, popup instance selected a unity instance ,Cannot attach .

I recorded my steps using the screen recording software!
can you using Mac OS system "Quick Time Player" or other support software play it video file

and , i get double unity instance, but i don't know who is who, can you show project name on item lable text ?

video file:
[test_New.mov.zip](https://github.com/Unity-Technologies/vscode-unity-debug/files/2162097/test_New.mov.zip)

**Comment by bujian** [comment]
I have the same problem ,The latest 2.6.3 works
Mine environment: Win10  Unity 5.6.0 VsCode 1.25.0

And I also have some suggestions:
1. Would it be better if the chosen UnityEditor process can be remembered so that I dont have to choose process again (if project name can be provided instead of process num, that‘s perfect)
2. New vscode Instance can be created when I open scripts from a new project instead of reopening from same instance (thats a habit from vs)
Are those reasonable and possible ?

**Comment by fjqingyou** [comment]
@miniwolf 

2.6.3 can attach! Thanks!

but , multiple target Unity instance selected item list show text Unfriendly.

for example
Unity Editor(Unity)(Process 1 Id)
Unity Editor(Unity)(Process 2 Id)
Unity Editor(Unity)(Process 3 Id)
Unity Editor(Unity)(Process 4 Id)
.......

if have multiple unity Instance , It is difficult to tell which is the target process, can you get unity instance open project name or project path?

**Comment by miniwolf** [comment]
Unfortunately no. The process id and name is the only thing that I have accessible. On newer editor versions, inside the library folder, a `EditorInstance.json` contains the information to bind process id and project name together.

**Comment by miniwolf** [comment]
@bujian
`Would it be better if the chosen UnityEditor process can be remembered so that I dont have to choose process again (if project name can be provided instead of process num, that‘s perfect)?`

How would this work? Would the project remember the process id? If you reopen the editor you would have to reopen this information. And if by some chance a new process has gotten this id we would try to attach to it and fail very badly. 


`New vscode Instance can be created when I open scripts from a new project instead of reopening from same instance (thats a habit from vs)
Are those reasonable and possible?`
It should be that each project folder is attached to one window instance. So when you open a file from the same project you will reuse the same window. Is this not the case with you?

**Comment by miniwolf** [comment]
@fjqingyou that sounds like a very dirty and low level operation. Getting the window, at least on windows, requires you to access functions like this: https://msdn.microsoft.com/en-us/library/ms633497(VS.85).aspx
and to retrieve the process ID from that window using this: https://msdn.microsoft.com/en-us/library/ms633498(v=vs.85).aspx
And then I would have to figure out how to do the same thing on linux and mac.

I am not saying it is not possible.

**Comment by fjqingyou** [comment]
Sorry

I'm using ProcessModule and ProcessThread on mac os platform try get Unity Process information ,  

Try The process to find the memory module Mono script compiled file Assembly-CSharp.dll

But failed, I did not find this module in the Unity process. unfortunately!

It’s really not feasible to experience it personally! Well, current!

**Comment by bujian** [comment]
@miniwolf 
yes, you are right.
I think I'd have to think more before questioning.It's really a technicial work.

**Comment by miniwolf** [comment]
We have been working on adding new information (project name for the Editor) to the attach command. This feature will be added in an upcoming release of this debugger extension.
